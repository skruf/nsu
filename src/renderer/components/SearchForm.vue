<style lang="stylus">
.search-form
  display flex
  margin 10px 0 20px
  .el-button
    margin-left 20px
</style>

<template>
  <el-form ref="form" class="search-form" @submit.prevent.native="submit">
    <el-input
      v-model="v"
      @input="input"
      @clear="submit"
      :placeholder="placeholder"
      :clearable="true"
    ></el-input>

    <el-button native-type="submit" type="primary">Search</el-button>
  </el-form>
</template>

<script>
export default {
  name: "SearchForm",
  props: {
    value: { type: String, default: "" },
    placeholder: { type: String, default: "Enter search words.." }
  },
  data: function() {
    return {
      v: this.value
    }
  },
  methods: {
    input() {
      this.$emit("input", this.v)
      if(this.v === "") {
        this.submit()
      }
    },
    submit() {
      this.$emit("submit", this.v)
    }
  }
}
</script>
