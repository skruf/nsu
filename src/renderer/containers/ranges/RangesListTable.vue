<style lang="stylus">
</style>

<template>
  <div class="ranges-list-table">
    <search-form
      v-model="rangesSearchFilter"
      placeholder="Search for a range by name or area"
      @submit="rangesSetSearchFilterAsync"
    />

    <div v-loading="rangesListIsLoading || rangesRemoveIsLoading">
      <el-table
        :data="rangesList"
        :sort-by="rangesSortBy"
        row-key="_id"
        class="table-clickable"
        empty-text
        @selection-change="rangesSelectionChange"
        @row-click="rangesRowClick"
        @sort-change="rangesSetSortingAsync"
      >
        <el-table-column
          type="selection"
          width="40"
        />

        <el-table-column
          prop="name"
          label="Name/Area"
          sortable="custom"
          :sort-orders="rangesSortOrders"
        >
          <template slot-scope="scope">
            <h6 class="h6">
              {{ scope.row.name }}
            </h6>
            <small class="small">
              {{ scope.row.area }}
            </small>
          </template>
        </el-table-column>

        <el-table-column
          prop="type"
          label="Type"
          sortable="custom"
          :sort-orders="rangesSortOrders"
        />

        <el-table-column
          prop="country"
          label="Country"
          sortable="custom"
          :sort-orders="rangesSortOrders"
        />

        <el-table-column
          prop="createdAt"
          label="Added On"
          sortable="custom"
          width="150px"
          :sort-orders="rangesSortOrders"
        >
          <template slot-scope="scope">
            {{ scope.row.createdAt | moment("MM.DD.YY") }}
          </template>
        </el-table-column>

        <el-table-column
          width="50"
          align="right"
        >
          <template slot="header">
            <div
              class="table-actions"
              :class="{ 'disabled': !rangesHasSelection }"
            >
              <el-dropdown
                trigger="click"
                @command="rangesTableDispatchActions"
              >
                <span class="el-dropdown-link">
                  <i class="table-button el-icon-more" />
                </span>
                <el-dropdown-menu slot="dropdown">
                  <el-dropdown-item command="removeSelection">
                    <i class="el-icon-delete el-icon--left" /> Remove selected
                  </el-dropdown-item>
                </el-dropdown-menu>
              </el-dropdown>
            </div>
          </template>
          <template slot-scope="scope">
            <el-dropdown
              trigger="click"
              @command="rangesTableRowDispatchActions"
            >
              <span class="el-dropdown-link">
                <i class="table-button el-icon-more" />
              </span>
              <el-dropdown-menu slot="dropdown">
                <!-- <el-dropdown-item :command="{ handler: 'edit', payload: scope.row }">
                  <i class="el-icon-edit el-icon--left"></i> Rediger
                </el-dropdown-item> -->
                <el-dropdown-item :command="{ handler: 'rangesDelete', payload: scope.row }">
                  <i class="el-icon-delete el-icon--left" /> Slett
                </el-dropdown-item>
              </el-dropdown-menu>
            </el-dropdown>
          </template>
        </el-table-column>

        <template slot="empty">
          No ranges yet.
          <el-button
            type="text"
            @click="rangesOpenCreateDialog"
          >
            Create new?
          </el-button>
        </template>
      </el-table>

      <el-pagination
        layout="total, sizes, prev, pager, next"
        :page-size="rangesPageSize"
        :current-page="rangesPageCurrent"
        :page-sizes="[ 15, 30, 45, 60 ]"
        :total="rangesCount"
        @size-change="rangesSetPageSizeAsync"
        @current-change="rangesSetPageCurrentAsync"
      />
    </div>
  </div>
</template>

<script>
import { mapActions, mapMutations, mapState } from "vuex"
import SearchForm from "@/components/SearchForm"

export default {
  name: "RangesListTable",

  components: {
    SearchForm
  },

  data: () => ({
    rangesSelection: [],
    rangesSortOrders: [ "descending", "ascending" ]
  }),

  computed: {
    ...mapState("ranges", {
      rangesListIsLoading: "listIsLoading",
      rangesRemoveIsLoading: "removeIsLoading",
      rangesSortBy: "sortBy",
      rangesPageSize: "pageSize",
      rangesPageCurrent: "pageCurrent",
      rangesCount: "count",
      rangesList: "list"
    }),
    rangesHasSelection() {
      return this.rangesSelection.length > 0
    },
    rangesSearchFilter: {
      get() { return this.$store.state.ranges.searchFilterValue },
      set(search) { this.rangesSetSearchFilter(search) }
    }
  },

  async created() {
    await this.rangesListAsync()
  },

  methods: {
    ...mapMutations("ranges", {
      rangesSetSearchFilter: "SET_SEARCH_FILTER"
    }),

    ...mapActions("ranges", {
      rangesListAsync: "listAsync",
      rangesRemoveAsync: "removeAsync",
      rangesSetSortingAsync: "setSortingAsync",
      rangesSetPageSizeAsync: "setPageSizeAsync",
      rangesSetPageCurrentAsync: "setPageCurrentAsync",
      rangesSetSearchFilterAsync: "setSearchFilterAsync"
    }),

    rangesOpenCreateDialog() {
      this.$emit("rangesOpenCreateDialog")
    },

    rangesTableRowDispatchActions({ handler, payload }) {
      switch(handler) {
        case "rangesDelete":
          this.rangesDeleteAsync(payload)
          break
      }
    },

    async rangesDeleteAsync(range) {
      await this.$confirm(
        `This will remove ${range.name} permanently. Continue?`,
        "Warning!", {
          confirmButtonText: "Yes, I am sure",
          cancelButtonText: "Cancel",
          customClass: "dangerous-confirmation",
          type: "warning"
        }
      )
      try {
        await this.rangesRemoveAsync(range)
        this.$notify({
          type: "success",
          title: "Success",
          message: `${range.name} was removed from the database`
        })
      } catch(e) {
        this.$notify({
          type: "error",
          title: "Oops!",
          message: e.message
        })
      }
    },

    rangesRowClick(range, e) {
      if(e.target.className.includes("table-button")) {
        return
      }
      this.$router.push(`/ranges/${range._id}`)
    },

    rangesSelectionChange(ranges) {
      this.rangesSelection = ranges
    },

    rangesTableDispatchActions({ handler, payload }) {}
  }
}
</script>
