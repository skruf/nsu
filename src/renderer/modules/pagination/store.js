const initialState = {
  pageSize: 15,
  pageCurrent: 1,
  sortBy: "_id",
  searchFilterValue: null,
  searchFilterFields: [ "title", "organizer" ]
}

const state = () => initialState

const mutations = {
  SET_SEARCH_FILTER(state, search) {
    state.searchFilterValue = search
    state.pageCurrent = 1
  },
  SET_SORTING(state, { prop, order }) {
    state.sortBy = prop
    if(order === "ascending") {
      state.sortBy = `-${state.sortBy}`
    }
  },
  SET_PAGE_SIZE(state, size) {
    state.pageSize = size
    state.pageCurrent = 1
  },
  SET_PAGE_CURRENT(state, pageCurrent) {
    state.pageCurrent = pageCurrent
  }
}

const actions = {
  async setSorting({ commit, dispatch }, options) {
    commit("SET_SORTING", options)
    // await dispatch("listAsync")
  },
  async setPageSize({ commit, dispatch }, pageSize) {
    commit("SET_PAGE_SIZE", pageSize)
    // await dispatch("listAsync")
  },
  async setPageCurrent({ commit, dispatch }, pageCurrent) {
    commit("SET_PAGE_CURRENT", pageCurrent)
    // await dispatch("listAsync")
  },
  resetView({ commit }) {
    commit("RESET_SEARCH_FILTER")
    commit("RESET_PAGINATION")
    commit("RESET_SORTING")
  }
}

const getters = {
  sortBy: ({ sortBy }) => sortBy,
  pageSize: ({ pageSize }) => pageSize,
  pageCurrent: ({ pageCurrent }) => pageCurrent
}

export const computeOptions = (state) => ({
  limit: state.pageSize,
  skip: (state.pageCurrent - 1) * state.pageSize,
  sort: state.sortBy,
  search: {
    fields: state.searchFilterFields,
    value: state.searchFilterValue || null
  }
})

export default {
  namespaced: true,
  state,
  mutations,
  actions,
  getters
}
