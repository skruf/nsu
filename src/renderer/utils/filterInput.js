export default (item, stub) => {
  let data

  if(Array.isArray(item)) {
    data = item.map((i) => filter(i, stub))
  } else {
    data = filter(item, stub)
  }

  return data
}

const filter = (item, stub) => {
  const data = {}

  for(let key in item) {
    if(stub.hasOwnProperty(key)) {
      let value = item[key]
      if(Array.isArray(value)) {
        data[key] = value.map(
          (v) => typeof v === "object" ? filter(v, stub[key]) : v
        )
      } else if(typeof value === "object") {
        data[key] = value = filter(value, stub[key])
      } else {
        data[key] = value
      }
    }
  }

  return data
}
