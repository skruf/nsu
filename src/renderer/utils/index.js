export { default as setId } from "./setId"
export { default as setTimestamps } from "./setTimestamps"
export { default as queryHelper } from "./queryHelper"
export { default as tests } from "./tests"
