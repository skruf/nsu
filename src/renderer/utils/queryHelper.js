export default (state) => ({
  limit: state.pageSize,
  skip: (state.pageCurrent - 1) * state.pageSize,
  sort: state.sortBy,
  search: {
    fields: state.searchFilterFields,
    value: state.searchFilterValue || null
  }
})
