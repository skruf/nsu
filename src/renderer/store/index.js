import Vue from "vue"
import Vuex from "vuex"

import events from "./modules/events"
import participants from "./modules/events/participants"
import divisions from "./modules/events/divisions"
import contestants from "./modules/events/divisions/contestants"
import clubs from "./modules/clubs"
import members from "./modules/clubs/members"
import ranges from "./modules/ranges"
import classes from "./modules/classes"

Vue.config.devtools = true
Vue.use(Vuex)

const modules = {}

modules.events = events
modules.events.modules = { participants, divisions }
modules.events.modules.divisions.modules = { contestants }
modules.clubs = clubs
modules.clubs.modules = { members }
modules.ranges = ranges
modules.classes = classes

const store = new Vuex.Store({
  strict: process.env.NODE_ENV !== "production",
  modules: modules
})

export default store
