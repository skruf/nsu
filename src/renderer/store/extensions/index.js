import merge from "lodash.merge"
import pagination from "./pagination"

const extenders = {
  pagination
}

export default (store, extensions = []) => {
  extensions.forEach((extension) => {
    const extender = extenders[extension.module]
    const extended = extender(extension)
    store = merge(extended, store)
  })

  return store
}
