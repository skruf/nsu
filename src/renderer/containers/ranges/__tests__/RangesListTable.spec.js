import { tests } from "@/utils"
import { shallowMount } from "@vue/test-utils"
import Vuex from "vuex"
import {
  Table,
  TableColumn,
  Dropdown,
  DropdownMenu,
  DropdownItem,
  Button,
  Pagination,
  Loading
} from "element-ui"
import SearchForm from "../../../components/SearchForm"
import RangesListTable from "../RangesListTable"

import { rangesFixture } from "@/fixtures"
import store from "../../../store/modules/ranges"
jest.mock("../../../store/modules/ranges")

const localVue = tests.createVue([
  Vuex,
  Table,
  TableColumn,
  Dropdown,
  DropdownMenu,
  DropdownItem,
  Button,
  Pagination,
  Loading.directive
])

describe("RangesListTable", () => {
  const build = () => {
    return shallowMount(RangesListTable, {
      localVue,
      store: new Vuex.Store({
        strict: true,
        modules: { ranges: store }
      })
    })
  }

  beforeEach(() => {
    jest.resetAllMocks()
  })

  it("renders the component", () => {
    const wrapper = build()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should have a search form, table and pagination", () => {
    const wrapper = build()
    expect(wrapper.find(SearchForm).exists()).toBe(true)
    expect(wrapper.find("el-table-stub").exists()).toBe(true)
    expect(wrapper.find("el-pagination-stub").exists()).toBe(true)
  })

  it.only("should get a list of ranges", () => {
    const wrapper = build()
    expect(store.actions.listAsync).toHaveBeenCalled()
    expect(wrapper.vm.rangesList).toBe(rangesFixture)
  })

  it("should filter table by searching for a name", () => {
    const wrapper = build()
    wrapper.vm.rangesSetSearchFilter(rangesFixture[0].name)
    console.log(wrapper.vm.rangesSearchFilter)
  })
})
