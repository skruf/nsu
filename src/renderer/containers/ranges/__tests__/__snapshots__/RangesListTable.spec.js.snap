// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangesListTable renders the component 1`] = `
<div class="ranges-list-table">
  <search-form-stub value="" placeholder="Search for a range by name or area"></search-form-stub>
  <div>
    <el-table-stub data="[object Object],[object Object]" fit="true" rowkey="_id" showheader="true" emptytext="" selectonindeterminate="true" class="table-clickable">
      <el-table-column-stub type="selection" width="40" resizable="true" filtermultiple="true" sortorders="ascending,descending,"></el-table-column-stub>
      <el-table-column-stub type="default" label="Name/Area" prop="name" sortable="custom" resizable="true" filtermultiple="true" sortorders="descending,ascending"></el-table-column-stub>
      <el-table-column-stub type="default" label="Type" prop="type" sortable="custom" resizable="true" filtermultiple="true" sortorders="descending,ascending"></el-table-column-stub>
      <el-table-column-stub type="default" label="Country" prop="country" sortable="custom" resizable="true" filtermultiple="true" sortorders="descending,ascending"></el-table-column-stub>
      <el-table-column-stub type="default" label="Added On" prop="createdAt" width="150px" sortable="custom" resizable="true" filtermultiple="true" sortorders="descending,ascending"></el-table-column-stub>
      <el-table-column-stub type="default" width="50" resizable="true" align="right" filtermultiple="true" sortorders="ascending,descending,"><template></template></el-table-column-stub> <template></template>
    </el-table-stub>
    <el-pagination-stub pagesize="10" pagercount="7" currentpage="1" layout="total, sizes, prev, pager, next" pagesizes="15,30,45,60"></el-pagination-stub>
  </div>
</div>
`;
